public/hooks/use_get_chunks_from_http_response.tsx

```diff
@@ -0,0 +1,162 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject } from 'rxjs';
+import { LLMResponseType, useChatState } from './use_chat_state';
+import { streamDeserializer } from '../../common/utils/stream/serializer';
+import { HttpResponse } from '../../../../src/core/public';
+import { SendResponse, StreamChunk } from '../../common/types/chat_saved_object_attributes';
+import { MessageContentPool } from '../utils/message_content_pool';
+
+export const useGetChunksFromHTTPResponse = () => {
+  const { chatStateDispatch } = useChatState();
+
+  const getConsumedChunk$FromHttpResponse = async (props: {
+    fetchResponse: HttpResponse<ReadableStream | Record<string, unknown>>;
+    abortController: AbortController;
+  }) => {
+    const chunk$ = new BehaviorSubject<StreamChunk | undefined>(undefined);
+    const messageContentPool = new MessageContentPool();
+    props.abortController.signal.addEventListener('abort', () => {
+      messageContentPool.stop();
+    });
+    if (props.fetchResponse.body?.getReader) {
+      chatStateDispatch({
+        type: 'updateResponseType',
+        payload: {
+          type: LLMResponseType.STREAMING,
+        },
+      });
+      const reader = (props.fetchResponse.body as ReadableStream).getReader();
+
+      const decoder = new TextDecoder();
+
+      async function processNextChunk() {
+        try {
+          if (props.abortController.signal.aborted) {
+            // If the abort controller is aborted.
+            // stop reading the stream.
+            return;
+          }
+          const { done, value } = await reader.read();
+
+          if (done) {
+            messageContentPool.inputComplete();
+            return;
+          }
+
+          const chunk = decoder.decode(value);
+
+          try {
+            const chunkObjects = streamDeserializer(chunk);
+            for (const chunkObject of chunkObjects) {
+              if (chunkObject.event === 'appendMessage') {
+                messageContentPool.addMessageContent(
+                  chunkObject.data.messageId,
+                  chunkObject.data.content
+                );
+              } else {
+                chunk$.next(chunkObject);
+              }
+            }
+          } catch (e) {
+            chunk$.error(e);
+            return;
+          }
+
+          processNextChunk();
+        } catch (error) {
+          chunk$.error(error);
+        }
+      }
+
+      processNextChunk();
+    } else {
+      chatStateDispatch({
+        type: 'updateResponseType',
+        payload: {
+          type: LLMResponseType.TEXT,
+        },
+      });
+      const response = (await props.fetchResponse.body) as Partial<SendResponse>;
+      chunk$.next({
+        event: 'metadata',
+        data: response,
+      });
+
+      // Complete right after next will only eat the emit value
+      // add a setTimeout to delay it into next event loop
+      setTimeout(() => {
+        messageContentPool.inputComplete();
```

Comment:
Why we need to call `messageContentPool.inputComplete`  here? For pure text response, seems we don't need to append message to the message pool.

Reply:
Removed since we refactor this hook.